File: app/src/main/java/com/hossameid/blescanner/presentation/BluetoothViewModel.java
package com.hossameid.blescanner.presentation;

import android.annotation.SuppressLint;
import android.app.Application;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothManager;
import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanFilter;
import android.bluetooth.le.ScanResult;
import android.bluetooth.le.ScanSettings;
import android.content.Intent;
import android.os.Handler;
import android.util.Log;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;

import com.hossameid.blescanner.system.MyBleForegroundService;

import java.util.List;

public class BluetoothViewModel extends AndroidViewModel {
    private final android.bluetooth.le.BluetoothLeScanner bluetoothLeScanner;
    private boolean scanning = false;
    private boolean deviceFound = false;
    private final Handler handler = new Handler();
    private final Runnable stopScanRunnable;

    // Stops scanning after 10 seconds.
    private static final long SCAN_PERIOD = 10000;

    private final MutableLiveData<String> scanResult = new MutableLiveData<>();
    private BluetoothDevice device = null;

    @SuppressLint("MissingPermission")
    public BluetoothViewModel(Application application) {
        super(application);
        BluetoothManager bluetoothManager = application.getSystemService(BluetoothManager.class);
        android.bluetooth.BluetoothAdapter bluetoothAdapter = bluetoothManager.getAdapter();
        bluetoothLeScanner = bluetoothAdapter.getBluetoothLeScanner();

        scanResult.setValue("not scanning");

        stopScanRunnable = () -> {
            scanning = false;
            bluetoothLeScanner.stopScan(leScanCallback);

            if (deviceFound)
                scanResult.setValue("device found");
            else
                scanResult.setValue("device not found");

            Log.d("BLE_SCAN_CALLBACK", "scanLeDevice: stopped scanning");
        };
    }

    @SuppressLint("MissingPermission")
    public void scanLeDevice(String macAddress, String name) {
        if (!scanning) {
            // Stops scanning after a predefined scan period.
            handler.postDelayed(stopScanRunnable, SCAN_PERIOD);

            scanning = true;
            deviceFound = false;
            ScanFilter.Builder builder = new ScanFilter.Builder();

            if (!macAddress.isEmpty())
                builder.setDeviceAddress(macAddress);

            if (!name.isEmpty())
                builder.setDeviceName(name);

            ScanFilter filter = builder.build();

            ScanSettings settings = new ScanSettings.Builder()
                    .setCallbackType(ScanSettings.CALLBACK_TYPE_ALL_MATCHES)
                    .build();

            bluetoothLeScanner.startScan(List.of(filter), settings, leScanCallback);
            scanResult.setValue("Scanning");
            Log.d("BLE_SCAN_CALLBACK", "scanLeDevice: started scanning");
        } else {
            scanning = false;
            bluetoothLeScanner.stopScan(leScanCallback);
        }
    }

    // Device scan callback.
    private final ScanCallback leScanCallback = new ScanCallback() {
        @SuppressLint("MissingPermission")
        @Override
        public void onScanResult(int callbackType, ScanResult result) {
            super.onScanResult(callbackType, result);
            Log.d("BLE_SCAN_CALLBACK", "onScanResult: " + result);

            deviceFound = true;
            device = result.getDevice();
            scanResult.setValue("device found");
            handler.removeCallbacks(stopScanRunnable);
            scanning = false;
            bluetoothLeScanner.stopScan(leScanCallback);

            Log.d("BLE_SCAN_CALLBACK", "scanLeDevice: stopped scanning");
        }
    };

    public MutableLiveData<String> getScanResult() {
        return scanResult;
    }

    public BluetoothDevice getDevice() {
        return device;
    }
}

File: app/src/main/java/com/hossameid/blescanner/presentation/BluetoothViewModel.kt
package com.hossameid.blescanner.presentation

import android.annotation.SuppressLint
import android.app.Application
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothManager
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.bluetooth.le.ScanSettings.CALLBACK_TYPE_ALL_MATCHES
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout

class BluetoothViewModel(application: Application) : AndroidViewModel(application) {
    private val bluetoothManager = application.getSystemService(BluetoothManager::class.java)
    private val bluetoothAdapter = bluetoothManager.adapter
    private val bluetoothLeScanner = bluetoothAdapter.bluetoothLeScanner
    private var scanning = false

    // Stops scanning after 10 seconds.
    private val SCAN_PERIOD: Long = 10000

    @SuppressLint("MissingPermission")
    fun scanLeDevice(macAddress: String, name: String) {
        viewModelScope.launch(Dispatchers.Main){
            if (!scanning) {
                // Start scanning with a timeout
                try {
                    scanning = true

                    val filter = ScanFilter.Builder().setDeviceName(name).setDeviceAddress(macAddress).build()
                    val settings = ScanSettings.Builder().setCallbackType(CALLBACK_TYPE_ALL_MATCHES).build()
                    bluetoothLeScanner.startScan(mutableListOf(filter), settings, leScanCallback)

                    // Stops scanning after a pre-defined scan period using withTimeout
                    withTimeout(SCAN_PERIOD) {
                        // The scanning will continue until the timeout occurs
                        delay(SCAN_PERIOD) // This is just to keep the coroutine active
                    }
                } catch (e: TimeoutCancellationException) {
                    // Timeout occurred, stop scanning
                } finally {
                    scanning = false
                    bluetoothLeScanner.stopScan(leScanCallback)
                }
            } else {
                scanning = false
                bluetoothLeScanner.stopScan(leScanCallback)
            }
        }
    }

    // Device scan callback.
    private val leScanCallback: ScanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            super.onScanResult(callbackType, result)
            Log.d("BLE SCAN", "onScanResult: $result")
        }
    }
}